code/obj/item/gun/energy.dm

```diff
@@ -216,25 +216,116 @@ TYPEINFO(/obj/item/gun/energy)
 
 
 ////////////////////////////////////// Antique laser gun
-// Part of a mini-quest thing (see displaycase.dm). Typically, the gun's properties (cell, projectile)
-// won't be the default ones specified here, it's here to make it admin-spawnable (Convair880).
-/obj/item/gun/energy/laser_gun/antique
+TYPEINFO(/obj/item/gun/energy/antique)
+	mats = 0
+/obj/item/gun/energy/antique
+	HELP_MESSAGE_OVERRIDE("You can use a <b>screwdriver</b> to open or close the maintenance panel. While the panel is open, you can insert lens and small coil to upgrade the weapon.")
 	name = "antique laser gun"
 	icon_state = "caplaser"
 	item_state = "capgun"
-	desc = "It's a kit model of the Mod.00 'Lunaport Legend' laser gun from Super! Protector Friend. With realistic sound fx and exciting LED display! This one has been hazardously upgraded."
+	cell_type = /obj/item/ammo/power_cell/tiny
+	force = 7
+	desc = "It's a kit model of the Mod.00 'Lunaport Legend' laser gun from Super! Protector Friend. With realistic sound fx and exciting LED display!"
 	muzzle_flash = "muzzle_flash_laser"
-	cell_type = null
 	uses_charge_overlay = TRUE
 	charge_icon_state = "caplaser"
 
-	New()
-		if (!src.current_projectile)
-			src.set_current_projectile(new /datum/projectile/laser/glitter)
-		if (isnull(src.projectiles))
-			src.projectiles = list(src.current_projectile)
-		..()
-		src.UpdateIcon()
+	var/obj/item/coil/small/myCoil = null
+	var/obj/item/lens/myLens = null
+	var/panelOpen = FALSE
+
+	examine(mob/user)
+		. = ..()
+		if(src.panelOpen)
+			. += "The maintenance panel is open."
+
+	attackby(obj/item/item, mob/user)
+		. = ..()
+		if(isscrewingtool(item))
+			user.show_text("You [src.panelOpen ? "close" : "open"] the maintenance panel.", "blue")
+			src.panelOpen = !src.panelOpen
+			if(!src.panelOpen)
+				if(src.determineProjectiles() >= 3)//highest tier
+					user.unlock_medal("Tinkerer", 1)
+		if(istype(item, /obj/item/coil/small))
+			if(panelOpen)
+				user.show_text("You insert [item]", "blue")
+				user.drop_item(item)
+				if(src.myCoil)
+					user.put_in_hand_or_drop(src.myCoil)
+				src.myCoil = item
+				item.set_loc(src)
+			else
+				user.show_text("You need to unscrew the maintenance panel first!", "red")
+		if (istype(item, /obj/item/lens))
+			if(panelOpen)
+				user.show_text("You insert [item]", "blue")
+				user.drop_item(item)
+				if(src.myLens)
+					user.put_in_hand_or_drop(src.myLens)
+				src.myLens = item
+				item.set_loc(src)
+			else
+				user.show_text("You need to unscrew the maintenance panel first!", "red")
+
+	canshoot(mob/user)
+		//configures the projectiles and makes sure it can actually shoot
+		if(src.determineProjectiles() < 1)
```

Comment:
This probably shouldn't be running the entire evaluation proc every time the gun fires, why not just evaluate it when the parts are swapped out?

Reply:
The hypothetical in my mind was if someone put something that would destroy itself in there, like a burning material. Since I didnt implement it exposing the parts to heat it probably isnt a very common scenario though. Moreso it's just convenient because youd need to check that it actually has all the parts anyways.

Reply:
Moved the quality analysis call to happen when you screw the maintenance panel closed. It also now cannot be fired while the maintenance panel is open.

Reply:
> The hypothetical in my mind was if someone put something that would destroy itself in there, like a burning material. Since I didnt implement it exposing the parts to heat it probably isnt a very common scenario though. Moreso it's just convenient because youd need to check that it actually has all the parts anyways.

One option you have there is to add an `Exited()` override and additionally recalculate whenever an atom leaves the contents of the gun